pkg/restore/log_client.go

```diff
@@ -0,0 +1,1008 @@
+	for tableID, name := range l.meta.Names {
+		schema, table := ParseQuoteName(name)
+		if !l.tableFilter.MatchTable(schema, table) {
+			log.Info("filter tables",
+				zap.String("schema", schema),
+				zap.String("table", table),
+				zap.Int64("tableID", tableID),
+			)
+			continue
+		}
+		tableIDs = append(tableIDs, tableID)
+	}
+
+	for _, tID := range tableIDs {
+		tableID := tID
+		// FIXME update log meta logic here
+		dir := fmt.Sprintf("%s%d", tableLogPrefix, tableID)
+		opt := &storage.WalkOption{
+			SubDir:    dir,
+			ListCount: -1,
+		}
+		if ok, err := l.restoreClient.storage.FileExists(ctx, opt.SubDir); !ok {
+			// this could happen after duplicate create drop table
+			log.Debug("path not exists", zap.String("path", opt.SubDir), zap.Error(err))
+			continue
+		}
+		err := l.restoreClient.storage.WalkDir(ctx, opt, func(path string, size int64) error {
+			fileName := filepath.Base(path)
+			shouldRestore, err := l.needRestoreRowChange(fileName)
+			if err != nil {
+				return err
+			}
+			if shouldRestore {
+				if _, ok := rowChangeFiles[tableID]; ok {
+					rowChangeFiles[tableID] = append(rowChangeFiles[tableID], path)
+				} else {
+					rowChangeFiles[tableID] = []string{path}
+				}
+			}
+			return nil
+		})
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// sort file in order
+	for tID, files := range rowChangeFiles {
+		sortFiles := files
+		sort.Slice(sortFiles, func(i, j int) bool {
+			if filepath.Base(sortFiles[j]) == logPrefix {
+				return true
+			}
+			return sortFiles[i] < sortFiles[j]
+		})
+		rowChangeFiles[tID] = sortFiles
+	}
+
+	return rowChangeFiles, nil
+}
+
+func (l *LogClient) writeToTiKV(ctx context.Context, kvs kv.Pairs, region *RegionInfo) ([]*sst.SSTMeta, error) {
+	firstKey := codec.EncodeBytes([]byte{}, kvs[0].Key)
+	lastKey := codec.EncodeBytes([]byte{}, kvs[len(kvs)-1].Key)
+
+	meta := &sst.SSTMeta{
+		Uuid:        uuid.NewV4().Bytes(),
+		RegionId:    region.Region.GetId(),
+		RegionEpoch: region.Region.GetRegionEpoch(),
+		Range: &sst.Range{
+			Start: firstKey,
+			End:   lastKey,
+		},
+	}
+
+	leaderID := region.Leader.GetId()
+	clients := make([]sst.ImportSST_WriteClient, 0, len(region.Region.GetPeers()))
+	requests := make([]*sst.WriteRequest, 0, len(region.Region.GetPeers()))
+	for _, peer := range region.Region.GetPeers() {
+		cli, err := l.importerClient.GetImportClient(ctx, peer.StoreId)
+		if err != nil {
+			return nil, err
+		}
+
+		wstream, err := cli.Write(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		// Bind uuid for this write request
+		req := &sst.WriteRequest{
+			Chunk: &sst.WriteRequest_Meta{
+				Meta: meta,
+			},
+		}
+		if err = wstream.Send(req); err != nil {
+			return nil, err
+		}
+		req.Chunk = &sst.WriteRequest_Batch{
+			Batch: &sst.WriteBatch{
+				// FIXME we should discuss about the commit ts
+				// 1. give a batch of kv a specify commit ts
+				// 2. give each kv the backup commit ts
+				CommitTs: oracle.ComposeTS(time.Now().Unix()*1000, 0),
```

Comment:
🤔 this mean every tikv peer will receive a different commit TS? how about move the calculation out of the loop?

Reply:
sure, I'll move it out of loop, to make every peer receive the same commit ts.